### ISSUE TITLE
Docker Swarm node starts routing requests to wrong containers

### ISSUE THREAD

**Issue by rkeilty** [docker/docker#25985]
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:22:43 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 33
 Running: 3
 Paused: 0
 Stopped: 30
Images: 15
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 277
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay null host
Swarm: active
 NodeID: 16e15nh1rf84uf6k5h6czxmp9
 Is Manager: true
 ClusterID: 9du6skrzpon6x3dfv8kdh5ygh
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.0.0.161
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-68-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.337 GiB
Name: ip-10-0-0-161
ID: AZKE:RARU:VGAZ:R6VQ:FWGS:DPNC:REID:QM2P:XRNO:CEL3:7K2C:FV2P
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: **********
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running a two-node swarm on AWS.  1 managers, 1 workers.

**Steps to reproduce the issue:**
1. There are 2 distinct web services being run in the swarm.  One service (let's call it `foo`) runs it's application on port 3031, the other (`bar`) on 4000.  Each binds to that same port on the host machine.  Things get created without issue `replicas=5` so there are a mix of services running on both nodes.
2. After some time, requests intended for `foo` are actually getting routed to `bar`.  Even though the two are running on different container ports, and running on different exposed swarm ports.

**Describe the results you received:**
I would expect that requests to swarm for port 4000 would route to `bar`, and 3031 would route to a `foo` container.

**Describe the results you expected:**
After some time, requests intended for `foo` are actually getting routed to `bar`.  Even though the two are running on different container ports, and running on different exposed swarm ports.

**Additional information you deem important (e.g. issue happens only occasionally):**
This only happens after some time (meaning that anywhere from 5-10 _updates_ to that service over 2-30 hours have occurred to ensure new images are being pushed into the wild.

Both services are attached to the `ingress` network, they do not have any other networks attached.

This tends to only happen to a isolated nodes in the swarm, not across the entire swarm.  Also it always seems to happen to a specific container thats being routed to _on that node_.  For example, the problem always seems to lead to the following diagnosis:

* Attempt to hit `foo` (3031) using cURL on node B (`curl http://127.0.0.1:3031/foo-request`)
* Each request to this node seems to result in being load balanced to a container for `bar` (logs clearly show a `bar` container on node B getting the request for `http://127.0.0.1:3031/foo-request`
* This node that is routing `foo` requests to `bar` definitely has `foo` services up and running.

`docker service rm` and then creating again, and letting it sit, seems to get things going for a short period of time, but things always get "confused" over time on nodes.

The swarm logs show this message occasionally, but don't correlate with the issue: `time="2016-08-22T08:33:35.667874882Z" level=warning msg="2016/08/22 08:33:35 [WARN] memberlist: Was able to reach ip-10-0-0-161 via TCP but not UDP, network may be misconfigured and not allowing bidirectional UDP\n"`

**Is there a way to look at the routing table being used on a node for the ingress network?**  I can't seem to find a way to just look at the routing table to see if routes are not being updated after a period of time.

**Comment by cpuguy83** [comment]
ping @mrjana @sanimej

**Comment by fphilippon** [comment]
Hello,

I was able to reproduce the issue using an overlay swarm network. Pretty annoying :(

**Comment by mnganesh** [comment]
Hello,

Even i am facing this issue.

**Comment by mnganesh** [comment]
~$ uname -a
Linux ip-10-224-128-100 3.13.0-91-generic #138-Ubuntu SMP Fri Jun 24 17:00:34 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mnganesh** [comment]
Hello,

Even i have been facing this same issue lately. After ~30 hours of swarm run without restart this starts showing up. Removing the service temporarily fixes it. Long time fix seems to be complete swarm restart.
`~$ uname -a
Linux ip-10-224-128-100 3.13.0-91-generic #138-Ubuntu SMP Fri Jun 24 17:00:34 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

`~$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 18
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: active
 NodeID: 4sd2bi6sbz038o4oinlxkhzbt
 Is Manager: true
 ClusterID: dgtj020aqy3wl3fln45h17nex
 Managers: 1
 Nodes: 5
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.224.128.100
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-91-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859 GiB
Name: ip-10-224-128-100
ID: YKWQ:EIOL:662H:DHUQ:KWGI:6R6Z:GSZ5:6Q2Y:5TXY:44O6:RSG3:EPGG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8`

**Comment by fphilippon** [comment]
Hi,
I did exactly the same test using  4.4.19 version of the linux kernel and i couldn't reproduce this strange behavior.
It seems this bug only apply to systems having an "old" kernel in the 3.x branch.

**Comment by fphilippon** [comment]
Please forget what i said before.
The issue came back after few hours of services running.

**Comment by mu5h3r** [comment]
Hi!
I have the same problem. 
Looks like the old ip addresses are sometimes not removed from the virtual services. It is happening after a few hours of uptime and several services updates.
Restarting of the servers resolves this problem, but not for long.

```
root@srv-docker-manager01:~# uname -a
Linux srv-docker-manager01 4.4.0-36-generic #55-Ubuntu SMP Thu Aug 11 18:01:55 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

```
root@srv-docker-manager01:~# docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

Script to create services:
```
#!/bin/bash
SERVICE_BASE_NAME=nginx-test
for SERVICE_ID in 1 2 3; do
         SERVICE_NAME=${SERVICE_BASE_NAME}-${SERVICE_ID}
         docker service create --name ${SERVICE_NAME} -p 3111${SERVICE_ID}/tcp nginx
done
```

Script to update services:
```
#!/bin/bash
SERVICE_BASE_NAME=nginx-test
VERSION=`date -u +%s`
for SERVICE_ID in 1 2 3; do
        SERVICE_NAME=${SERVICE_BASE_NAME}-${SERVICE_ID}
        docker service update --container-label-add version=${VERSION} ${SERVICE_NAME}
done
```

Docker service ps output (after a few services updates):
```
ID                         NAME                IMAGE  NODE                 DESIRED STATE  CURRENT STATE           ERROR
f3l609p1ygl91kf8z20i5gz5k  nginx-test-1.1      nginx  srv-docker-worker04  Running        Running 2 seconds ago
8nucf3kr12qz6oe7mevxru3l7   \_ nginx-test-1.1  nginx  srv-docker-worker04  Shutdown       Shutdown 3 seconds ago
9xcf2oib7uuvidjo9vaygv8u6   \_ nginx-test-1.1  nginx  srv-docker-worker04  Shutdown       Shutdown 3 minutes ago
dk7rxoax6vyh3lqd7jspnuc80   \_ nginx-test-1.1  nginx  srv-docker-worker04  Shutdown       Shutdown 6 minutes ago
dc9mtmewt3l6rmrzr9f9hg4g0   \_ nginx-test-1.1  nginx  srv-docker-worker04  Shutdown       Shutdown 7 minutes ago
ID                         NAME                IMAGE  NODE                 DESIRED STATE  CURRENT STATE           ERROR
6fq62x9zt4ai4ojorwpfmy8h5  nginx-test-2.1      nginx  srv-docker-worker04  Running        Running 1 seconds ago
50otd86kt8fwswvgizwqwmazx   \_ nginx-test-2.1  nginx  srv-docker-worker04  Shutdown       Shutdown 3 seconds ago
7t0g92nc95ctagoijn6ipem4b   \_ nginx-test-2.1  nginx  srv-docker-worker04  Shutdown       Shutdown 3 minutes ago
0qicvme4y8563f13h988ap124   \_ nginx-test-2.1  nginx  srv-docker-worker04  Shutdown       Shutdown 6 minutes ago
9t96f87eedoobopio1l4jo789   \_ nginx-test-2.1  nginx  srv-docker-worker04  Shutdown       Shutdown 7 minutes ago
ID                         NAME                IMAGE  NODE                 DESIRED STATE  CURRENT STATE           ERROR
3js7wil0ns5snt632m4wju86r  nginx-test-3.1      nginx  srv-docker-worker01  Running        Running 2 seconds ago
dhqdfeh4to7jiq6tmjsrux3si   \_ nginx-test-3.1  nginx  srv-docker-worker04  Shutdown       Shutdown 3 seconds ago
3wq7n6g337od7rtn8mzisph36   \_ nginx-test-3.1  nginx  srv-docker-worker01  Shutdown       Shutdown 3 minutes ago
77w6pjo1hy3ix8kbg41puaeyz   \_ nginx-test-3.1  nginx  srv-docker-worker04  Shutdown       Shutdown 6 minutes ago
0840oc5pg1gv44ytdgys52st5   \_ nginx-test-3.1  nginx  srv-docker-worker01  Shutdown       Shutdown 7 minutes ago  
```
Iptables rules:
```
0     0 MARK       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:31111 MARK set 0x12e
0     0 MARK       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:31112 MARK set 0x12f
0     0 MARK       tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:31113 MARK set 0x130
```

Ipvsadm output:
```
FWM  302 rr
  -> 10.255.0.30:0                Masq    1      0          0
  -> 10.255.0.35:0                Masq    1      0          0
  -> 10.255.0.36:0                Masq    1      0          0
FWM  303 rr
  -> 10.255.0.32:0                Masq    1      0          0
  -> 10.255.0.35:0                Masq    1      0          0
  -> 10.255.0.36:0                Masq    1      0          0
FWM  304 rr
  -> 10.255.0.34:0                Masq    1      0          0
  -> 10.255.0.37:0                Masq    1      0          0
```

```
root@srv-docker-worker01:~# docker network inspect ingress
[
    {
        "Name": "ingress",
        "Id": "bwyyhmlymgjvkmufw73x4wzek",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.255.0.0/16",
                    "Gateway": "10.255.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "b704dbe2170e7f7075153524f7034e0dfd6b309fefcb176a5b01eafd75870145": {
                "Name": "nginx-test-3.1.3js7wil0ns5snt632m4wju86r",
                "EndpointID": "0e344b4553a85a28ae291d4224a543c4d8303cb37abaae6478ee6ee1acff9abd",
                "MacAddress": "02:42:0a:ff:00:25",
                "IPv4Address": "10.255.0.37/16",
                "IPv6Address": ""
            },
            "ingress-sbox": {
                "Name": "ingress-endpoint",
                "EndpointID": "755a8931989fc4ea1232aabd8e3a182b8ad542940d52dbfc38f87296d1a99e7d",
                "MacAddress": "02:42:0a:ff:00:02",
                "IPv4Address": "10.255.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "256"
        },
        "Labels": {
            "com.docker.swarm.internal": "true"
        }
    }
]
```

```
root@srv-docker-worker04#
docker network inspect ingress
[
    {
        "Name": "ingress",
        "Id": "bwyyhmlymgjvkmufw73x4wzek",
        "Scope": "swarm",
        "Driver": "overlay",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.255.0.0/16",
                    "Gateway": "10.255.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "57247fd866e8a03a44cae32ce8ced1604ec3465b8eb46e26be034cb71b68004d": {
                "Name": "nginx-test-2.1.6fq62x9zt4ai4ojorwpfmy8h5",
                "EndpointID": "5587edc9620a83c9fdee2f71d8022832c1ab728c1b52cad1958ce712e6515918",
                "MacAddress": "02:42:0a:ff:00:24",
                "IPv4Address": "10.255.0.36/16",
                "IPv6Address": ""
            },
            "ca854a0d58f1a35310a17e8d22c63e4272ebba6dede93411ea92ec3fd8f7f474": {
                "Name": "nginx-test-1.1.f3l609p1ygl91kf8z20i5gz5k",
                "EndpointID": "4400369de830c391d1eca6d93198de90b45b4d2ac8eb22501a946c47f788eda8",
                "MacAddress": "02:42:0a:ff:00:23",
                "IPv4Address": "10.255.0.35/16",
                "IPv6Address": ""
            },
            "ingress-sbox": {
                "Name": "ingress-endpoint",
                "EndpointID": "be95570f940cfc1c6b9004991e40021de7bfbb93f4f755f055b6984f8750c1d8",
                "MacAddress": "02:42:0a:ff:00:0b",
                "IPv4Address": "10.255.0.11/16",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.driver.overlay.vxlanid_list": "256"
        },
        "Labels": {}
    }
]
```

**Comment by mrjana** [comment]
@rkeilty @mnganesh @fphilippon @mu5h3r Is there anything happening in the cluster that triggers this problem when you say this problem happens after some time? i.e is there any service getting scaled up or down or getting updated for image or other parameters or even if a service is getting removed or added back or nodes restarted? If any of you can post the daemon logs of the node where wrong routing is happening that would be helpful.

**Comment by fphilippon** [comment]
On our side this problem happens without any action. Like said before it seems always affect containers in one node in the swarm.

**Comment by freewilll** [comment]
We have this problem too.  We not only get requests going to incorrect containers but also "No route to host" errors. In one example case, the ipvsadm output contained 3 IP addresses that didn't belong to any containers and belonging to a real container, resulting in 3 out of 4 requests producing a no route to host error.

**Comment by freewilll** [comment]
Looks potentially related to #25325

**Comment by freewilll** [comment]
Not sure if this helps, but here's a couple of points and things we noticed.

- Swarm cluster: 4 nodes
- Swarm version: 1.12.1 kernel version: 4.4.0-36-generic
- We run 5 replicas by default of each service. Things were failing without us changing the number of replicas manually (read on)
- Our setup is heavily used, with new services being created on a regular basis.
- Newly spawned services tend to behave fine for a while
- After we scaled all services down to one replica, I examined the kernel's virtual server table on the four nodes. They had a different number of entries, ranging from one IP to four IP addresses. Only one was reachable, so doing a curl to the container via the load balancer on the nodes lead to different percentages of failures.

**Comment by mu5h3r** [comment]
Looks like we have some network errors, may be this is causing the problem.
[dockerd.txt](https://github.com/docker/docker/files/462101/dockerd.txt)

**Comment by LK4D4** [comment]
ping @mrjana 
Some logs with errors here.

**Comment by mrjana** [comment]
@mu5h3r Thanks for the logs. That does confirm my suspicion that the gossip is indeed timing out in your case. It could be a network issue. I will study the logs a bit more and try simulating a transient network issue to see how the cluster recovers.

**Comment by dangra** [comment]
I was seeing the same problem in 5 nodes swarm cluster, then reproduced it in single node swarm. Reported (probably duplicate) as #26590.

**Comment by mrjana** [comment]
@mu5h3r On simulating a network partitioning scenario and see how the cluster recovers I found issues in the gossip cluster recovery and pushed a fix here docker/libnetwork#1446 which should fix the recovery part.

@dangra I don't think what you are experiencing in #26590 is the same issue as this.

**Comment by vieux** [comment]
This issue is a duplicate of #26563
